### ISSUE TITLE
task 2.0

### ISSUE THREAD

**Issue by JohnKhandygo** [ADsty/shedule#6]
**Общая формулировка**

Разработать консольную утилиту, реализующую указанную в задании функциональность, на языках Java и/или Kotlin. При проектировании использовать объектно-ориентированный подход. Разработать автоматические тесты для отдельных частей программы. Написанный код держать в отдельном проекте или модуле в репозитории на GitHub.

**uniq**

Объединение последовательностей одинаковых идущих подряд строк в файле в одну:
file задаёт имя входного файла. Если параметр отсутствует, следует считывать текст с консоли.
Флаг -o ofile  задаёт имя выходного файла. Если параметр отсутствует, следует выводить результаты на консоль.
Флаг -i означает, что при сравнении строк следует не учитывать регистр символов.
Флаг -s N означает, что при сравнении строк следует игнорировать первые N символов каждой строки. Выводить нужно первую строку.
Флаг -u означает, что следует выводить в качестве результата только уникальные строки.
Флаг -с означает, что перед каждой строкой вывода следует вывести количество строк, которые были заменены данной (т.е. если во входных данных было 2 одинаковые строки, в выходных данных должна быть одна с префиксом “2”).

Command line: uniq [-i] [-u] [-c] [-s num] [-o ofile] [file]

В случае, когда какое-нибудь из имён файлов указано неверно, следует выдать ошибку.

Кроме самой программы, следует написать автоматические тесты к ней.

**Comment by JohnKhandygo** [comment]
Добавил в задание пункт про скрипты сборки/запуска.

Если будете делать это задание в другом репозитории, скиньте сюда ссылку.

**Comment by JohnKhandygo** [comment]
Не забудьте написать, когда задание будет готово для проверки.

**Comment by JohnKhandygo** [comment]
Первое задание зачтено. 

Рекомендуется следующий подход к разработке второго задания:
1. Реализовать класс со всей необходимой логикой так, как Вы это делали в предыдущем задании (с покрытием тестами и прочим). Другими словами, забудьте на время, что Ваше приложение должно работать из консоли.
2. Отдельно написать "обертку", которая будет парсить аргументы командной строки (используя библотеку) и вызывать разработанные на первом этапе методы. Скорее всего то, что будет написано на данном этапе даже не потребуется покрывать тестами, поскольку функциональность сторонних библотек обычно не тестируется (либо тестируется в каких-то очень сложных случаях). 


Собственно после готовности первого этапа можно начинать проверку.

**Comment by ADsty** [comment]
Здравствуйте , программа почти готова , были проблемы  с флагами  , но понял что можно вынести отдельно флаг -s N и флаг -i. У меня есть вопрос - если я использую в своей программе сторонний класс FileLineIterator , но он был добавлен не в качестве библиотеки , а как один из классов моей программы , то можно ли его использовать вообще , не будет ли это плагиатом и нужно ли указать сайт , откуда я его взял?

**Comment by JohnKhandygo** [comment]
Если Вы изначально нашли класс в виде исходного кода (типа на stackoverflow или на каком-нибудь другом форуме), то лучше в javadoc указать, откуда он взят просто для удобства проверки. Если Вы выдрали этот класс из библотеки, то лучше все-таки саму библиотеку использовать. 

А чем Вас не устроил обычный способ чтения строк из файла? 
```
try (BufferedReader br = new BufferedReader(new FileReader(file))) {
    String line;
    while ((line = br.readLine()) != null) {
       // process the line.
    }
}
```

**Comment by ADsty** [comment]
Мне в реализации нужно сравнивать соседние строки , пока читал материал по buffered reader не заметил там метода как в итераторе .next() , который бы вызвал следующую строку, и решил поискать на форумах , вот и нашел FileLineIterator. Вполне возможно , что в будущем уберу его , просто читая про метод .readLine() я не совсем понял , работает ли он как в итераторе .next() . То есть если мне нужно будет сравнить две соседние строки , смогу ли я просто написать if(line.equals(br.readLine())) или нет? На форумах про все это как то размыто написано , вот и остаются вопросы

**Comment by JohnKhandygo** [comment]
Окей. Давайте сначала посмотрим на то, что у Вас получилось. В принципе ничего криминального в этом нет.

**Comment by ADsty** [comment]
Что-то я поспешил с тем что программа почти готова) Она работает , но нужно править ее логику , ибо пока что если флаг сравнивает строки без учета регистра , то выводит он все строки в нижнем регистре. И подскажите пожалуйста как заставить Buffered Writer переносить строку , когда я записываю их в файл

**Comment by JohnKhandygo** [comment]
https://docs.oracle.com/javase/8/docs/api/java/io/BufferedWriter.html#newLine--

**Comment by JohnKhandygo** [comment]
Напишите тогда еще раз, когда можно будет проверять.

**Comment by ADsty** [comment]
Я решил загрузить неправильный вариант , ибо уже голову себе сломал. Флаги работали все , за исключением флага -u , ввод с консоли и вывод на консоль работают . С флагом -u какие то проблемы и из-за этого я не могу продолжить работу . Возможно я не вижу какую то простую вещь , которую нужно исправить. Репозиторий назван так же как и задание uniq

**Comment by ADsty** [comment]
Тестировал все я вручную создавая файлы на компьютере

**Comment by JohnKhandygo** [comment]
Окей, я посмотрю, но даже с первого взгляда понятно, что немножко бред написан. Такого тестирования совершенно точно недостаточно. Подумаю, как натолкнуть Вас на путь истинный.

**Comment by ADsty** [comment]
Да , это довольно странный метод тестирования , но мне почему то захотелось посмотреть , что же будет если я реальные файлы с моего компьютера , которые не находятся в идее , запущу в мою программу и увидел , что получалось если я буду писать разные флаги , и у меня они все работали как нужно , но когда я решил запустить вообще все флаги и в файл записать несколько подряд идущих одинаковых строк и одну отличную от них , я ожидал что выведет две строки , но он вывел только одну и так же было если я добавлял еще одну уникальную строку - он опять выводил только одну строку. И я уже просто всю программу проглядел и не понимаю почему он не хочет выводить все уникальные строки. У меня стоит в самом начале цикла условие что если строка j не равна строке h , которая идет за строкой j , то записывать строку  p , которая является абсолютно такой же строкой j , просто без изменений , которые были сделаны флагами. И программа почему то не хочет ее записывать

**Comment by ADsty** [comment]
Возможно обычные тесты раскроют мне глаза на мои ошибки , но я на тот момент очень смутно представлял как мне их делать , но сейчас мне пришло в голову несколько идей , как протестировать программу

**Comment by ADsty** [comment]
В принципе разобрался в чем была проблема , все дело в копировании чужого класса - итератор выводил значение переменной string и затем менял его , а мне казалось что он сначала меняет значение переменной string и потом выводит его , получается что у меня j и h были одной и той же строчкой

**Comment by ADsty** [comment]
Вроде бы смог починить все. Пока что в программе только один тест , который проверят метод work() со всеми флагами одновременно , что наверное говорит о том , что они теперь работают все . Можно ли делать тесты только под разные конфигурации флагов и используя только метод work , так как по сути только он и выполняет всю работу и взаимодействует со всеми остальными методами , или же нужно для каждого метода отдельный тест как в прошлый раз?

**Comment by JohnKhandygo** [comment]
Нет, прямо все методы покрывать не надо. В принципе надо просто показать, что программа работает корректно при любых значениях флагов. В Вашем случае это сложно потому что комбинаций много, постарайтесь покрыть наиболее важные из них. Совершенно точно должно быть показано, что программа работает с консольным вводом/выводом. У Вас по коду с этим проблемы.

По коду: есть несколько грубых ошибок, но с точки зрения логики обработки строк все вроде бы окей. Более точные замечания постараюсь в субботу/воскресенье написать.

**Comment by ADsty** [comment]
Ну вроде как я не из файла читаю строки, просто я искал на форумах способ с консоли в файл записывать строки , там постоянно светилась строчка вроде этой :  BufferedReader reader = new BufferedReader(new InputStreamReader(System.in)); . Как я понял она читает строки с консоли , но если это не так то я переправлю программу

**Comment by JohnKhandygo** [comment]
Перенес описание задания в новый репозиторий. Все замечания будут появляться там.
